src/core/lib/security/credentials/ssl/ssl_credentials.cc

```diff
@@ -66,12 +68,40 @@ grpc_ssl_credentials::create_security_connector(
   absl::optional<std::string> overridden_target_name =
       args->GetOwnedString(GRPC_SSL_TARGET_NAME_OVERRIDE_ARG);
   auto* ssl_session_cache = args->GetObject<tsi::SslSessionLRUCache>();
+  tsi_ssl_session_cache* tsi_ssl_session_cache =
+      ssl_session_cache == nullptr ? nullptr : ssl_session_cache->c_ptr();
+
+  const tsi_ssl_root_certs_store* root_store;
+  // Use default root certificates.
+  if (config_.pem_root_certs == nullptr) {
+    // TODO(gtcooke94) make sure to remove this memory
+    const char* pem_root_certs =
+        grpc_core::DefaultSslRootStore::GetPemRootCerts();
+    if (pem_root_certs == nullptr) {
+      gpr_log(GPR_ERROR, "Could not get default pem root certs.");
+      return nullptr;
+    }
+    size_t root_len = strlen(pem_root_certs);
+    char* default_roots = strcpy(new char[root_len + 1], pem_root_certs);
+    config_.pem_root_certs = default_roots;
+    root_store = grpc_core::DefaultSslRootStore::GetRootStore();
+  } else {
+    config_.pem_root_certs = config_.pem_root_certs;
+    root_store = nullptr;
+  }
+
+  grpc_security_status status = initialize_client_handshaker_factory(
+      &config_, config_.pem_root_certs, root_store, tsi_ssl_session_cache);
```

Comment:
I don't think we can pass in the `tsi_ssl_session_cache` in this way because it is coming from the ChannelArgs, which are tied to the channel rather than to the SSL credentials.

To get around this, I was thinking that we could add a new `tsi_ssl_session_cache*` parameter to the `tsi_ssl_client_handshaker_factory_create_handshaker` API, and plumb the session cache pointer from here to where the security connector calls that API. WDYT?

If we can do that, then I think we can also move all of this logic to the constructor of the SSL credentials.

Reply:
I think that makes sense with the `tsi_ssl_session_cache`. `tsi_ssl_client_handshaker_factory_create_handshaker` ends up getting called here - https://github.com/grpc/grpc/blob/4295f0250c6344c1e492bdb45d4ad01ae28d9177/src/core/lib/security/security_connector/ssl/ssl_security_connector.cc#L140 and we have access to the channel args, so that should be easy to plumb.

I don't know if we can move all the logic to the SSL credentials - we're still making a lot of calls to OpenSSL that can fail when creating and populating the SSL_CTX, so we need to be able to do a status return